architecture/crds.md

```diff
@@ -0,0 +1,324 @@
+2. It is impossible to pre-verify a chart that has both a CRD and CRs for that CRD
+3. `helm delete` becomes a very, very dangerous command, as it can destabilize a cluster by deleting a CRD
+4. `helm upgrade` and `helm rollback` can have surprising consequences that are difficult to find and debug
+
+We still haven't solved #2, but by adding some special treatment of CRDs in Helm 3, we eliminated #3-4 and fixed #1.
+
+> **IMPORTANT**: It is still possible to use Helm this way. For CRD/Operator-only charts, this is a perfectly legitimate way of packaging.
+
+### Chart Authors Have Control
+
+In this category, the assertion is that the _chart author_ should be able to declare the conditions under which a chart is created, upgrade, or deleted. For example, one chart author may stipulate that on an upgrade, their chart should avoid modifying the CRD. Another chart author, though, may choose that on upgrade, their chart will modify the existing CRD.
+
+The problems that this approach must address are:
+
+- How can a chart author know when it is safe to install, upgrade, or delete a CRD without harming the broader cluster?
+- How can massive data loss be prevented when a CRD is deleted?
+- How can two versions of the same chart (with different CRD versions) be safely installed on the same cluster?
+- Can the user opt to override this behavior if they detect an issue? (And if so, how?)
+
+As a broad category, this solution has been proposed several times. In fact, Helm 3 has a very weak version of this: If a CRD is stored in the `crds/` folder, by default it will be installed (and a user can skip with `--skip-crds`). This install only occurs if the CRD is not already present. If the CRD is present _it is not modified_. Following this strategy, _at worst_ the only package that sustains damage if the CRD is not at the right version is the package just installed. No existing behavior is broken.
+
+We frequently hear the complain that the above is not enough; that we _must_ also support updates and deletions. But no _specific comprehensive_ solution has surfaced that solves the problems above. Adherents of this view typically claim that users should merely accept the fact that a chart operation may destabilize an entire cluster by altering or deleting CRDs. We do not view that as an acceptable risk.
+
+### Helm Users Have Control
+
+In this model, the creation, modification, and deletion of CRDs is under the explicit control of the Helm user. A chart may declare a CRD, but that CRD is not changed unless the user specifically requests it.
+
+One could imagine, in this model, the following commands and flags:
+
+```console
+$ helm install --with-crds ...
+$ helm upgrade --allow-crd-modification ...
+$ helm delete --unsafe-delete-crds ...
+```
+
+But with this model, there are also problems:
+
+- How does the user even know when a chart has a CRD? (If the answer is "the user has to read the chart," this is a non-starter)
```

Comment:
Helm is in the position to help notify the user of this. for example, on delete it could state that it previously managed crds.